Source/DependencyInversion/TenantScopedProvidersBuilder.cs

```diff
@@ -64,4 +67,57 @@ IServiceProvider CreateTenantContainer(TenantId tenant)
         services.AddSingleton(tenant);
         return _factory(_serviceProvider, tenant, services);
     }
+
+    static void DiscoverTenantScopedServices(IServiceCollection services)
+    {
+        ForAllAllScannedAssemblies(assembly =>
+        {
+            foreach (var type in GetExportedTypes(assembly))
+            {
+                var attribute = type.GetCustomAttribute<PerTenantAttribute>();
+                if (attribute is null)
+                {
+                    continue;
+                }
+                var implementors = type.GetInterfaces().Where(_ => _ != typeof(IDisposable)).ToList();
+                if (attribute.RegisterAsSelf)
+                {
+                    implementors.Add(type);
+                }
+
+                foreach (var serviceType in implementors)
+                {
+                    services.Add(new ServiceDescriptor(serviceType, type, attribute.Lifetime));
+                }
+            }
+        });
+    }
+
+    static IEnumerable<Type> GetExportedTypes(Assembly assembly)
+    {
+        try
+        {
+            return assembly.ExportedTypes;
+        }
+        catch
+        {
+            return Enumerable.Empty<Type>();
```

Comment:
Log on failure?

Reply:
We have pretty much this same code somewhere else as well. When it fails it's basically because the assembly could not be loaded, which probably in 99% of the cases is not interesting.

Reply:
If we would have logging on these I would rather have the assembly scanning be cached or centarlized in some way and logged as Trace or something